New function to replace to_python.Excel (xlwings)

lyralan opened the pull request:
1. Updated `to_python.Excel` Class with `openpyxl`-based approach, keeping methods and arguments the same, leaving deprecation message where needed;
2. Updated `i_o.py` and test files accordingly;
3. Tested new `to_python.Excel` on Adapter and other tools.
4. Updated `__init__.py` and `setup.py`.
5. Formatted.

lyralan commented:
This PR is applicable to reading and loading inputs. I just realized that LCC still uses xlwings to write outputs because unlike Shipper and NIA which use `adapter.io` (where I made commits), it imports `adapter.comm.excel` instead, which will again ask permission to access output files/folders. So far only LCC has this issue. If we want to fully step out from xlwings, changes need to be made for writing functions in LCC as well.

evaneill commented:
Thanks Lyra!

Do you have access to be able to branch this repo? It's not necessary to fork for our purposes. This is also working for me on our repos, though. I'm very excited to move away from xlwings!

One big suggestion is to consider backwards compatibility, since it's really hard to tell how much would break if we replace the `Excel` class from `to_python` with some functions. It *does* look like as long as only `i_o.IO` is used it should be acting the same way, which is good.

I'd say it'd be better if you update the guts of the old `Excel` class using these new openpyxl tools, keeping the method names etc the same, and supporting kwargs where possible (where not, adding deprecation message). Maybe these functions can be static methods too. Without looking at the old class in detail I don't know if this is mostly a 1-1 method replacement, or more piecemeal. Maybe Milica and Zheng also have thoughts on this.

One much smaller thing is that you can also update the version in `adapter.__init__`, probably the minor version (so update to 1.3.0)

lyralan commented:
> 

Hi Evan - Yes, I have access to this repo. I forked the branch because back then I didn't have permission. 

So far I haven't had any backward compatibility issues. It does look like `to_python.Excel` is only used in `adapter.i_o`, but happy to wrap `openpyxl`-based functions into a new `Excel` class for consistency. 

Sure, I'll update `__init__.py` accordingly.

0xd5dc reviewed (commented):
I fixed one trivial conflict for you. Please fix the rest.

lyralan commented:
> I fixed one trivial conflict for you. Please fix the rest.

O I also just fixed all conflicts and pushed. But somehow `test_to_python.TestDb` failed (which is not relevant to where I made changes).

lyralan commented:
> I fixed one trivial conflict for you. Please fix the rest.

Did `test_to_python.TestDb.test_load_df` pass on your end? 

`(self.good_db.load(table_names=['table3'])['table3'])` [line 89](https://github.com/lyralan/Adapter/blob/fc7c5d702d546b17fefdf07c17a4c37577f7f200/adapter/tests/test_to_python.py#L86) would throw an KeyError on my end.

Also, [line 72](https://github.com/lyralan/Adapter/blob/fc7c5d702d546b17fefdf07c17a4c37577f7f200/adapter/tests/test_to_python.py#L72) seems to be always true. At least for me, the code below would always return 0.

```
len((self.good_db.load(table_names=['table1', 'table2','table3'])).keys()- {'table1', 'table2'})
len((self.good_db.load(table_names=['table1', 'table2','table3'])).keys()- {'table1'})
len((self.good_db.load(table_names=['table1', 'table2'])).keys()- {'table1'})
len((self.good_db.load(table_names=['table2'])).keys()- {'table1'})
len((self.good_db.load(table_names=[])).keys()- {'table1'})
```

![Screen Shot 2022-01-14 at 1 35 54 PM](https://user-images.githubusercontent.com/22651612/149588433-1574fa67-0043-498a-883f-d27c3ae5400e.png)

lyralan commented:
> > I fixed one trivial conflict for you. Please fix the rest.
> 
> Did `test_to_python.TestDb.test_load_df` pass on your end?
> 
> `(self.good_db.load(table_names=['table3'])['table3'])` [line 89](https://github.com/lyralan/Adapter/blob/fc7c5d702d546b17fefdf07c17a4c37577f7f200/adapter/tests/test_to_python.py#L86) would throw an KeyError on my end.
> 
> Also, [line 72](https://github.com/lyralan/Adapter/blob/fc7c5d702d546b17fefdf07c17a4c37577f7f200/adapter/tests/test_to_python.py#L72) seems to be always true. At least for me, the code below would always return 0.
> 
> ```
> len((self.good_db.load(table_names=['table1', 'table2','table3'])).keys()- {'table1', 'table2'})
> len((self.good_db.load(table_names=['table1', 'table2','table3'])).keys()- {'table1'})
> len((self.good_db.load(table_names=['table1', 'table2'])).keys()- {'table1'})
> len((self.good_db.load(table_names=['table2'])).keys()- {'table1'})
> len((self.good_db.load(table_names=[])).keys()- {'table1'})
> ```
> 
> ![Screen Shot 2022-01-14 at 1 35 54 PM](https://user-images.githubusercontent.com/22651612/149588433-1574fa67-0043-498a-883f-d27c3ae5400e.png)

@0xd5dc @milicag Okay both, I found the problem in the master branch.

First, `test_to_python.TestDb` would not raise an error (yet it should) when loading `corrupt.db` because `raise IOError` was commented out in [commit1](https://github.com/lyralan/Adapter/commit/55a4a900ddbaf4d5751ad337087f34b42ca98034). We should uncomment the related code, otherwise `test_to_python` would fail.

Second, `to_python.TestDb` loads test files by 
```
self.db = Db(file_path="./test.db"
```
which would fail the tests unless the tests are run from the `adapter/tests` directory. To pass the tests, I would need to run 
```
cd adapter/tests
python -m unittest test_to_python
```
rather than `README.md`'s instructions on how to run the unit tests
```
python -m unittest adapter.tests.test_to_python
```
The older version before [commit2](https://github.com/lyralan/Adapter/commit/beefd2144f4abb7b5fe379292f9525f35d009740) loads test files by 

```
self.db = Db(r"adapter/tests/test.db")
```
which works for me and is consistent with `README.md`.

With that, further changes in `to_python.py` and `test_to_python.py` are needed. Do you want me to create another PR or just commit in this PR?

evaneill commented:
Thanks for all the updates

1. For the expected IOError in the test, I'd actually go look at an older commit (like v1.1.0) and see if we used to throw errors when no tables were loaded (I dont remember if we did or not). Whatever we did then is probably what we'd want to keep doing, and can implement in this PR if it's as simple as either altering a test or uncommenting lines
2. For the file paths in the tests, that can probably just be a new issue. IMO you suggested the best option: `os.path.join(__file__,"test.db")`, so that adapter tests could be run anywhere

lyralan commented:
> Thanks for all the updates
> 
> 1. For the expected IOError in the test, I'd actually go look at an older commit (like v1.1.0) and see if we used to throw errors when no tables were loaded (I dont remember if we did or not). Whatever we did then is probably what we'd want to keep doing, and can implement in this PR if it's as simple as either altering a test or uncommenting lines
> 2. For the file paths in the tests, that can probably just be a new issue. IMO you suggested the best option: `os.path.join(__file__,"test.db")`, so that adapter tests could be run anywhere

Thanks Evan! The old commit would throw an error when loading an empty file (i.e. corrupt.db). 

If John and Milica also agree on this, I'll just make a commit in this PR.

milicag commented on adapter/to_python.py:
```diff
@@ -30,189 +65,105 @@ class Excel(object):
+        return dict_of_dfs
 
-            try:
+    def get_named_data_objects(self, data_object_name):
```
I would much rather have this as a part of `excel` module with this refactor. What are your thoughts on this?

milicag commented on adapter/to_python.py:
```diff
@@ -30,189 +65,105 @@ class Excel(object):
-                msg = "Failed to read input tables from {}."
+            data_object_name: str
+                worksheet reference to find
+                a regtangular region (cell or tuple of cells)
```
Typo ->`rectangular`

milicag reviewed (changes requested):
@lyralan Thank you for a great contribution. As @evaneill already hinted in his comment, I've addressed in my review aspects concerning the `Adapter` API. The review addressed topics such as the placement of methods, naming conventions, maintenance of of method arguments (in this case kwargs), as well as some aspects of code documentation style and logging. Please address the comments once you get a chance and I can take a look again.

lyralan commented on adapter/i_o.py:
```diff
@@ -221,14 +217,10 @@ def load(
-                if (qry_flags[file_path] is not None) and isinstance(
-                    table_names, str
-                ):
+                if (qry_flags[file_path] is not None) and isinstance(table_names, str):
```
Agreed and noted.

lyralan commented on adapter/tests/test_to_python.py:
```diff
@@ -1,29 +1,58 @@
 
 
-class ExcelTests(unittest.TestCase):
+class UtilTest(TestCase):
```
This test is for the general function `convert_data_object_to_df`, which is used in the `Excel` class.

milicag replied:
We prefer to stick to the naming convention where the test class contains the name of the class being tested, so please use `ExcelTests`.

lyralan commented on adapter/tests/test_to_python.py:
```diff
@@ -1,29 +1,58 @@
 
-    def tearDown(self) -> None:
-        del self.exl_loader
+    def test_convert_single_cell_to_df(self):
```
These two methods don't use the `Excel` class at all, and they are testing the util method `convert_data_object_to_df`. If we move it into the `ExcelTest` class, it will not be consistent with the `Excel` class.

milicag replied:
That's right, I am referring to `excel` module and the corresponding test module, rather than the `Excel` class from the `to_python` module.

milicag replied:
@lyralan Were you able to move the methods? Is it ready for me to I take another look?

lyralan commented:
@milicag I've addressed all your comments and updated Adapter accordingly.

evaneill commented on adapter/i_o.py:
```diff
@@ -453,9 +431,7 @@ def get_tables(
-        file_path = convert_network_drive_path(
-            file_path, mapping=self.os_mapping
-        )
+        file_path = convert_network_drive_path(file_path, mapping=self.os_mapping)
```
In addition, the `user_select_file` function also needs a case for linux. Updating for linux use can be a whole issue & PR, I made a trello card to try to describe it:

https://trello.com/c/PascRIFT/301-add-linux-support-for-file-prompt-and-convertnetworkpath

evaneill reviewed (approved):
Thanks for working on this & being responsive

milicag reviewed (approved):
@lyralan This tests ok on my end. Please merge and [apply a version tag](https://atcd.lbl.gov/source/wiki/versioning.html) when you get a chance. Thanks!

lyralan commented:
> @lyralan This tests ok on my end. Please merge and [apply a version tag](https://atcd.lbl.gov/source/wiki/versioning.html) when you get a chance. Thanks!

Just merged and tagged.
